File: prototype/snapshot.py
# prototype/snapshot.py — GameState 序列化/反序列化

import json
from prototype.terrain import Terrain
from prototype.game import GameState, init_game
from prototype.unit import Unit, City


def terrain_to_str(t: Terrain) -> str:
    return t.name


def str_to_terrain(s: str) -> Terrain:
    return Terrain[s]


def game_to_dict(gs: GameState) -> dict:
    """将 GameState 序列化为 JSON 兼容的 dict"""
    return {
        "version": "2026.07.01",
        "seed": gs.seed,
        "size": gs.size,
        "generator_id": gs.generator_id,
        "turn": gs.turn,
        "winner": gs.winner,
        "victory_type": gs.victory_type,
        "grid": [[{
            "terrain": terrain_to_str(c["terrain"]),
            "facility": {
                "type": c["facility"].facility_type,
                "player": c["facility"].player_id,
                "x": c["facility"].x,
                "y": c["facility"].y,
            } if c["facility"] else None
        } for c in row] for row in gs.grid],
        "units": [{
            "type": u.unit_type, "pid": u.player_id,
            "x": u.x, "y": u.y, "hp": u.hp,
            "atk": u.atk, "def": u.def_, "alive": u.alive,
        } for u in gs.units + gs.dead_units],
        "cities": [{
            "pid": c.player_id, "x": c.x, "y": c.y, "hp": c.hp,
        } for c in gs.cities],
        "economies": [{
            "food": e.food, "wood": e.wood, "gold": e.gold,
        } for e in gs.economies],
        "techs": [{
            "completed": list(t.completed),
            "researching": t.researching,
            "ticks": t.research_ticks,
        } for t in gs.techs],
        "action_log": gs.action_log,
    }


def game_to_json(gs: GameState, indent: int = 2) -> str:
    return json.dumps(game_to_dict(gs), indent=indent, ensure_ascii=False)


def dict_to_game(d: dict) -> GameState:
    """从 dict 重构 GameState（不含 action_log 重放——仅恢复静态状态）"""
    gs = GameState(seed=d["seed"], size=d["size"], generator_id=d["generator_id"])
    gs.turn = d["turn"]
    gs.winner = d["winner"]
    gs.victory_type = d["victory_type"]

    # grid
    gs.grid = [[{
        "terrain": str_to_terrain(c["terrain"]),
        "facility": None,
    } for c in row] for row in d["grid"]]
    # restore facilities
    for y, row in enumerate(d["grid"]):
        for x, c in enumerate(row):
            if c["facility"]:
                from prototype.unit import Facility
                gs.grid[y][x]["facility"] = Facility(
                    c["facility"]["type"],
                    c["facility"]["player"],
                    c["facility"]["x"],
                    c["facility"]["y"],
                )

    # units
    gs.units = []
    gs.dead_units = []
    for ud in d["units"]:
        u = Unit.create(ud["type"], ud["pid"], ud["x"], ud["y"])
        u.hp = ud["hp"]; u.atk = ud["atk"]; u.def_ = ud["def"]
        u.alive = ud["alive"]
        if u.alive:
            gs.units.append(u)
        else:
            gs.dead_units.append(u)

    # cities
    gs.cities = [City(c["pid"], c["x"], c["y"]) for c in d["cities"]]
    for i, c in enumerate(d["cities"]):
        gs.cities[i].hp = c["hp"]

    # economies
    from prototype.economy import Economy
    gs.economies = []
    for ed in d["economies"]:
        e = Economy(0)
        e.food = ed["food"]; e.wood = ed["wood"]; e.gold = ed["gold"]
        gs.economies.append(e)

    # techs
    from prototype.tech import TechManager
    gs.techs = []
    for td in d["techs"]:
        t = TechManager(0)
        t.completed = set(td["completed"])
        t.researching = td["researching"]
        t.research_ticks = td["ticks"]
        if "C3" in t.completed:
            t.has_academy = True
        gs.techs.append(t)

    gs.action_log = d.get("action_log", [])
    return gs


def json_to_game(s: str) -> GameState:
    return dict_to_game(json.loads(s))


# ─── GameReplay format ────────────────────────────────
# 轻量级回放格式：每回合只存变化量（units摘要 + economies + techs + events），
# 不存完整grid（地形不变）。用于 HTML 回放浏览器。

def snapshot_turn(gs) -> dict:
    """单回合状态快照——只存摘要，不存完整grid。
    返回 dict，符合 GameReplay turns[] 格式。
    """
    # Units summary
    units = []
    for u in gs.units:
        if u.alive:
            units.append({
                "type": u.unit_type,
                "pid": u.player_id,
                "x": u.x, "y": u.y,
                "hp": u.hp, "atk": u.atk, "def": u.def_,
            })

    # Cities
    cities = []
    for c in gs.cities:
        cities.append({
            "pid": c.player_id,
            "x": c.x, "y": c.y,
            "hp": c.hp,
        })

    # Economies
    economies = []
    for e in gs.economies:
        economies.append({
            "pid": e.pid,
            "food": e.food, "wood": e.wood, "gold": e.gold,
        })

    # Techs
    techs = []
    for t in gs.techs:
        techs.append({
            "pid": t.pid,
            "completed": sorted(list(t.completed)),
            "researching": t.researching,
            "research_ticks": t.research_ticks,
        })

    # Facilities count + positions
    from prototype.mapgen import get_facility
    facility_count = {0: 0, 1: 0}
    facilities = []  # list of {pid, type, x, y}
    for y in range(gs.size):
        for x in range(gs.size):
            f = get_facility(gs.grid, x, y)
            if f is not None:
                facility_count[f.player_id] += 1
                facilities.append({
                    "pid": f.player_id,
                    "type": f.facility_type,
                    "x": x, "y": y
                })

    # Events from action_log (last entry = this turn)
    events = []
    if gs.action_log:
        last = gs.action_log[-1]
        if last.get("turn") == gs.turn:
            # Parse action_log into events
            for pid_key, actions in [("p0", last.get("p0", [])), ("p1", last.get("p1", []))]:
                p = 0 if pid_key == "p0" else 1
                for act in actions:
                    atype = act.get("type", "")
                    if atype == "build":
                        u = _find_unit_in_snapshot(units, p, act)
                        if u:
                            events.append({
                                "type": "build",
                                "pid": p,
                                "x": u["x"], "y": u["y"],
                                "detail": "built facility",
                            })
                    elif atype == "move":
                        u = _find_unit_in_snapshot(units, p, act)
                        if u:
                            events.append({
                                "type": "move",
                                "pid": p,
                                "x": u["x"], "y": u["y"],
                                "detail": f"moved to ({u['x']},{u['y']})",
                            })
                    elif atype == "research" and act.get("tech_id"):
                        events.append({
                            "type": "research",
                            "pid": p,
                            "detail": f"started {act['tech_id']}",
                        })
                    elif atype == "produce_unit" and act.get("unit_type"):
                        events.append({
                            "type": "produce_unit",
                            "pid": p,
                            "detail": f"produced {act['unit_type']}",
                        })

    return {
        "turn": gs.turn,
        "units": units,
        "cities": cities,
        "economies": economies,
        "techs": techs,
        "facility_count": facility_count,
        "facilities": facilities,
        "events": events,
    }


def _find_unit_in_snapshot(units: list, pid: int, act: dict) -> dict | None:
    """从当前回合的 units snapshot 中找到执行动作的单位（基于 unit_idx）。"""
    ui = act.get("unit_idx", -1)
    # Count units belonging to pid
    pid_units = [u for u in units if u["pid"] == pid]
    if 0 <= ui < len(pid_units):
        return pid_units[ui]
    return None


def create_replay(gs, seed: int = 0, ai_a: str = "P0", ai_b: str = "P1") -> dict:
    """从已完成的 GameState 创建 GameReplay JSON。
    ai_a/ai_b: AI 名称（用于回放显示）。
    """
    turns = getattr(gs, "turn_snapshots", [])

    # Encode terrain grid (doesn't change during game, store once)
    terrain_grid = []
    from prototype.terrain import Terrain
    terrain_map = {Terrain.PLAIN: 0, Terrain.FOREST: 1, Terrain.MOUNTAIN: 2,
                   Terrain.WATER: 3, Terrain.CITY: 4}
    for y in range(gs.size):
        row = []
        for x in range(gs.size):
            t = gs.grid[y][x]["terrain"]
            row.append(terrain_map.get(t, 0))
        terrain_grid.append(row)

    return {
        "format_version": "1.0",
        "config": {
            "size": gs.size,
            "gen": gs.generator_id,
            "max_turns": 100,
            "seed": seed,
            "terrain_grid": terrain_grid,
            "ai_a": ai_a,
            "ai_b": ai_b,
        },
        "turns": turns,
        "result": {
            "winner": gs.winner,
            "victory_type": gs.victory_type,
            "final_turn": gs.turn,
        },
    }


def save_replay(gs, filepath: str, seed: int = 0):
    """保存 GameReplay JSON 到文件。"""
    replay = create_replay(gs, seed)
    with open(filepath, "w", encoding="utf-8") as f:
        json.dump(replay, f, indent=2, ensure_ascii=False)
    return replay

File: miniciv-core/src/snapshot.rs
// 单局回放输出 — 阶段 0.2(第三个 AI)
//
// 目的: 让人能"看一局到底发生了什么"。人类可观察性的最小前提,
//       也是引入人类锚点(验证金字塔第4层)的地基。
//
// 设计: snapshot_turn(gs) 从当前状态抓一帧; run_replay 驱动一局并逐回合收集。
//       原 create_replay 只拿到最终 gs 拿不到历史,故新增带记录的驱动器。

use crate::game::{GameState, init_game, step_game};
use crate::map::{Grid, Terrain};
use crate::ai::Agent;
use crate::constants::{MAP_W, MAP_H, MAX_TURNS};
use rand_chacha::ChaCha12Rng;
use rand::SeedableRng;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
pub struct GameReplay {
    pub format_version: String,
    pub config: ReplayConfig,
    pub turns: Vec<TurnSnapshot>,
    pub result: ReplayResult,
}

#[derive(Serialize, Deserialize)]
pub struct ReplayConfig {
    pub size: u8,
    pub gen: String,
    pub max_turns: u16,
    pub seed: u64,
    pub terrain_grid: Vec<Vec<u8>>,
    pub ai_a: String,
    pub ai_b: String,
    pub grid_type: String,
}

#[derive(Serialize, Deserialize)]
pub struct TurnSnapshot {
    pub turn: u16,
    pub units: Vec<UnitSnapshot>,
    pub cities: Vec<CitySnapshot>,
    pub economies: Vec<EconomySnapshot>,
    pub techs: Vec<TechSnapshot>,
    pub facility_count: std::collections::HashMap<u8, u8>,
    pub facilities: Vec<FacilitySnapshot>,
    pub events: Vec<EventSnapshot>,
}

#[derive(Serialize, Deserialize)] pub struct UnitSnapshot { pub unit_type: String, pub pid: u8, pub q: i32, pub r: i32, pub hp: i32, pub atk: i32, pub def: i32 }
#[derive(Serialize, Deserialize)] pub struct CitySnapshot { pub pid: u8, pub q: i32, pub r: i32, pub hp: i32 }
#[derive(Serialize, Deserialize)] pub struct EconomySnapshot { pub pid: u8, pub food: i32, pub wood: i32, pub gold: i32 }
#[derive(Serialize, Deserialize)] pub struct TechSnapshot { pub pid: u8, pub completed: Vec<String>, pub researching: Option<String>, pub research_ticks: u8 }
#[derive(Serialize, Deserialize)] pub struct FacilitySnapshot { pub pid: u8, pub facility_type: String, pub q: i32, pub r: i32 }
#[derive(Serialize, Deserialize)] pub struct EventSnapshot { pub event_type: String, pub pid: u8, pub detail: String }
#[derive(Serialize, Deserialize)] pub struct ReplayResult { pub winner: Option<u8>, pub victory_type: Option<String>, pub final_turn: u16 }

/// 地形 → u8(回放渲染用)
fn terrain_code(t: Terrain) -> u8 {
    match t {
        Terrain::Plain => 0,
        Terrain::Forest => 1,
        Terrain::Mountain => 2,
        Terrain::Water => 3,
        Terrain::City => 4,
    }
}

/// 从当前 GameState 抓一帧快照(纯函数,不改状态)。
pub fn snapshot_turn(gs: &GameState) -> TurnSnapshot {
    let mut units = Vec::new();
    for u in gs.units.iter().filter(|u| u.alive) {
        units.push(UnitSnapshot {
            unit_type: format!("{:?}", u.unit_type),
            pid: u.player_id,
            q: u.q, r: u.r, hp: u.hp,
            atk: u.atk, def: u.def,
        });
    }

    let cities = gs.cities.iter().map(|c| CitySnapshot {
        pid: c.player_id, q: c.q, r: c.r, hp: c.hp,
    }).collect();

    let economies = gs.economies.iter().enumerate().map(|(pid, e)| EconomySnapshot {
        pid: pid as u8, food: e.food, wood: e.wood, gold: e.gold,
    }).collect();

    let techs = gs.techs.iter().enumerate().map(|(pid, t)| {
        let mut completed: Vec<String> = t.completed.iter().cloned().collect();
        completed.sort();  // HashSet 无序 → 排序保证回放确定性
        TechSnapshot {
            pid: pid as u8,
            completed,
            researching: t.researching.clone(),
            research_ticks: t.research_ticks,
        }
    }).collect();

    // 扫全图收集设施
    let mut facilities = Vec::new();
    let mut facility_count = std::collections::HashMap::new();
    for r in 0..MAP_H as i32 {
        for q in 0..MAP_W as i32 {
            if let Some(f) = &gs.grid.get(q, r).facility {
                facilities.push(FacilitySnapshot {
                    pid: f.player_id,
                    facility_type: format!("{:?}", f.facility_type),
                    q, r,
                });
                *facility_count.entry(f.player_id).or_insert(0) += 1;
            }
        }
    }

    TurnSnapshot {
        turn: gs.turn,
        units, cities, economies, techs,
        facility_count, facilities,
        events: Vec::new(),  // 逐回合事件留待需要时补(当前胜利记在 result)
    }
}

fn terrain_grid(grid: &Grid) -> Vec<Vec<u8>> {
    (0..MAP_H as i32).map(|r| {
        (0..MAP_W as i32).map(|q| terrain_code(grid.get(q, r).terrain)).collect()
    }).collect()
}

/// 驱动一局并逐回合收集回放。P0 rng=seed, P1 rng=seed+1(与 eval 一致)。
pub fn run_replay(
    seed: u64, ai0: &dyn Agent, ai1: &dyn Agent,
    generator_id: &str, max_turns: u16,
) -> GameReplay {
    let mut gs = init_game(seed, generator_id);
    let mut rng0 = ChaCha12Rng::seed_from_u64(seed);
    let mut rng1 = ChaCha12Rng::seed_from_u64(seed + 1);

    let config = ReplayConfig {
        size: gs.size,
        gen: generator_id.to_string(),
        max_turns,
        seed,
        terrain_grid: terrain_grid(&gs.grid),
        ai_a: ai0.name().to_string(),
        ai_b: ai1.name().to_string(),
        grid_type: "hex".to_string(),
    };

    let mut turns = vec![snapshot_turn(&gs)];  // turn 0 初始帧
    while gs.winner.is_none() && gs.turn < max_turns {
        let a0 = ai0.decide(&gs, 0, &mut rng0);
        let a1 = ai1.decide(&gs, 1, &mut rng1);
        step_game(&mut gs, &a0, &a1);
        turns.push(snapshot_turn(&gs));
    }

    GameReplay {
        format_version: "1.0".to_string(),
        config,
        turns,
        result: ReplayResult {
            winner: gs.winner,
            victory_type: gs.victory_type.as_ref().map(|v| format!("{:?}", v)),
            final_turn: gs.turn,
        },
    }
}

/// 便捷:默认 MAX_TURNS。
pub fn run_replay_default(seed: u64, ai0: &dyn Agent, ai1: &dyn Agent, generator_id: &str) -> GameReplay {
    run_replay(seed, ai0, ai1, generator_id, MAX_TURNS)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::ai::fixed::BuilderAgent;
    use crate::ai::random::RandomAgent;

    #[test]
    fn test_回放帧数与结果一致() {
        let b = BuilderAgent;
        let r = RandomAgent;
        let rep = run_replay_default(50000, &b, &r, "balanced");
        // turns 含 turn0 初始帧 + 每回合一帧
        assert_eq!(rep.turns.len(), rep.result.final_turn as usize + 1);
        assert!(rep.result.winner.is_some());
        // 地图尺寸正确
        assert_eq!(rep.config.terrain_grid.len(), MAP_H as usize);
        assert_eq!(rep.config.terrain_grid[0].len(), MAP_W as usize);
    }

    #[test]
    fn test_回放确定性() {
        let b = BuilderAgent;
        let r = RandomAgent;
        let r1 = run_replay_default(50000, &b, &r, "balanced");
        let r2 = run_replay_default(50000, &b, &r, "balanced");
        assert_eq!(r1.result.final_turn, r2.result.final_turn);
        assert_eq!(r1.result.winner, r2.result.winner);
    }
}
